src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionService.cs

```diff
@@ -0,0 +1,667 @@
+                {
+                    initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: true));
+                }
+                else
+                {
+                    if (data.InitialTrigger.Reason == EditorCompletion.InitialTriggerReason.Deletion ||
+                        data.InitialTrigger.Reason == EditorCompletion.InitialTriggerReason.Invoke ||
+                        filterText.Length <= 1)
+                    {
+                        initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: false));
+                    }
+                }
+            }
+
+            if (initialListOfItemsToBeIncluded.Count == 0)
+            {
+                return Task.FromResult(HandleAllItemsFilteredOut(data.InitialTrigger.Reason, data.SelectedFilters, selectedFilters, mustSetSelection));
+            }
+
+            // If this was deletion, then we control the entire behavior of deletion
+            // ourselves.
+            if (data.InitialTrigger.Reason == EditorCompletion.InitialTriggerReason.Deletion)
+            {
+                return Task.FromResult(HandleDeletionTrigger(data.InitialSortedList, data.InitialTrigger.Reason, data.SelectedFilters, data.UpdateTrigger.Reason, filterText, initialListOfItemsToBeIncluded, mustSetSelection));
+            }
+
+            var caretPoint = session.TextView.GetCaretPoint(data.Snapshot.TextBuffer);
+            var caretPosition = caretPoint.HasValue ? caretPoint.Value.Position : (int?)null;
+
+            var snapshotForDocument = data.InitialSortedList.FirstOrDefault(i => i.Properties.ContainsProperty(CompletionItemSource.TriggerBuffer))?.Properties.GetProperty<ITextBuffer>(CompletionItemSource.TriggerBuffer).CurrentSnapshot ?? data.Snapshot;
+
+            return Task.FromResult(HandleNormalFiltering(
+                data.InitialSortedList,
+                snapshotForDocument,
+                caretPosition,
+                filterText,
+                data.SelectedFilters,
+                data.UpdateTrigger.Reason,
+                initialListOfItemsToBeIncluded,
+                data.InitialTrigger.Reason,
+                mustSetSelection));
+        }
+
+        private static RoslynTrigger GetRoslynTrigger(EditorCompletion.UpdateTrigger trigger)
+        {
+            RoslynTrigger roslynTrigger = default;
+            switch (trigger.Reason)
+            {
+                case EditorCompletion.UpdateTriggerReason.Insertion:
+                    roslynTrigger = RoslynTrigger.CreateInsertionTrigger(trigger.Character);
+                    break;
+                case EditorCompletion.UpdateTriggerReason.Deletion:
+                    roslynTrigger = RoslynTrigger.CreateDeletionTrigger(trigger.Character);
+                    break;
+            }
+
+            return roslynTrigger;
+        }
+
+        private EditorCompletion.FilteredCompletionModel HandleDeletionTrigger(
```

Comment:
Method name has _deletion_ and in first conditional you check _insertion_. Potential mismatch?